File: social-service/src/test/java/com/eirsteir/coffeewithme/social/repository/FriendshipRepositoryTest.java
package com.eirsteir.coffeewithme.social.repository;

import static org.assertj.core.api.Assertions.assertThat;

import com.eirsteir.coffeewithme.config.EventuateTestConfig;
import com.eirsteir.coffeewithme.social.domain.friendship.Friendship;
import com.eirsteir.coffeewithme.social.domain.friendship.FriendshipStatus;
import com.eirsteir.coffeewithme.social.domain.user.User;
import java.util.List;
import java.util.Optional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@DataJpaTest
@ActiveProfiles("test")
@Import(EventuateTestConfig.class)
@ExtendWith(SpringExtension.class)
class FriendshipRepositoryTest {
  private static final String REQUESTER_NICKNAME = "requester";
  private static final String ADDRESSEE_NICKNAME = "addressee";

  private User requester;
  private User addressee;

  @Autowired private TestEntityManager entityManager;

  @Autowired private FriendshipRepository friendshipRepository;

  @BeforeEach
  void setUp() {
    requester =
        entityManager.persistFlushFind(
            User.builder().id(1000L).nickname(REQUESTER_NICKNAME).build());

    addressee =
        entityManager.persistFlushFind(
            User.builder().id(1001L).nickname(ADDRESSEE_NICKNAME).build());

    entityManager.persistAndFlush(
        Friendship.builder()
            .requester(requester)
            .addressee(addressee)
            .status(FriendshipStatus.ACCEPTED)
            .build());
  }

  @Test
  void testFindByIdRequesterOrIdAddresseeAndStatusWhenNoResults() {
    List<Friendship> friendsFound =
        friendshipRepository.findByUserAndStatus(requester.getId(), FriendshipStatus.REQUESTED);

    assertThat(friendsFound).isEmpty();
  }

  @Test
  void testFindAllByExampleOfRequesterWhenRequesterIsAddressee() {
    User otherUser = entityManager.persistFlushFind(User.builder().id(1003L).build());

    entityManager.persistAndFlush(
        Friendship.builder()
            .requester(addressee)
            .addressee(requester)
            .status(FriendshipStatus.ACCEPTED)
            .build());

    List<Friendship> friendsFound =
        friendshipRepository.findByUserAndStatus(requester.getId(), FriendshipStatus.ACCEPTED);

    assertThat(friendsFound).hasSize(2);
  }

  @Test
  void testFindByIdRequesterIdAndIdAddresseeIdWhenExists() {
    Optional<Friendship> friendshipFound =
        friendshipRepository.findByIdRequesterIdAndIdAddresseeId(
            requester.getId(), addressee.getId());

    assertThat(friendshipFound).isPresent();
  }

  @Test
  void testFindByIdRequesterIdAndIdAddresseeIdWhenNotExists() {
    Optional<Friendship> friendshipFound =
        friendshipRepository.findByIdRequesterIdAndIdAddresseeId(requester.getId(), 100L);

    assertThat(friendshipFound).isEmpty();
  }
}

File: social-service/src/test/kotlin/com/eirsteir/coffeewithme/social/repository/FriendshipRepositoryTest.kt
package com.eirsteir.coffeewithme.social.repository

import com.eirsteir.coffeewithme.config.EventuateTestConfig
import com.eirsteir.coffeewithme.social.domain.friendship.Friendship
import com.eirsteir.coffeewithme.social.domain.friendship.FriendshipStatus
import com.eirsteir.coffeewithme.social.domain.user.User
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles
import java.util.*

@ActiveProfiles("test")
@Import(EventuateTestConfig::class)
@DataJpaTest
internal class FriendshipRepositoryTest {

    private lateinit var requester: User
    private lateinit var addressee: User

    @Autowired
    private lateinit var entityManager: TestEntityManager

    @Autowired
    private lateinit var friendshipRepository: FriendshipRepository

    @BeforeEach
    fun setUp() {
        requester = entityManager.persistFlushFind(
            User(id = 1000, nickname = REQUESTER_NICKNAME)
        )
        addressee = entityManager.persistFlushFind(
            User(id = 1001L, nickname = ADDRESSEE_NICKNAME)
        )
        entityManager.persistAndFlush(
            Friendship(
                requester = requester,
                addressee = addressee,
                status = FriendshipStatus.ACCEPTED,
                )
        )
    }

    @Test
    fun testFindByIdRequesterOrIdAddresseeAndStatusWhenNoResults() {
        val friendsFound = friendshipRepository.findByUserAndStatus(requester.id!!, FriendshipStatus.REQUESTED)

        assertThat(friendsFound).isEmpty()
    }

    @Test
    fun testFindAllByExampleOfRequesterWhenRequesterIsAddressee() {
        val otherUser = entityManager.persistFlushFind(User(id = 1003L))
        entityManager.persistAndFlush(
            Friendship(
                requester = addressee,
                addressee = requester,
                status = FriendshipStatus.ACCEPTED
            )
        )
        val friendsFound = friendshipRepository.findByUserAndStatus(requester.id!!, FriendshipStatus.ACCEPTED)

        assertThat(friendsFound).hasSize(2)
    }

    @Test
    fun testFindByIdRequesterIdAndIdAddresseeIdWhenExists() {
        val friendshipFound = friendshipRepository.findByIdRequesterIdAndIdAddresseeId(
            requester.id!!, addressee.id!!
        )

        assertThat(friendshipFound).isPresent
    }

    @Test
    fun testFindByIdRequesterIdAndIdAddresseeIdWhenNotExists() {
        val friendshipFound = friendshipRepository.findByIdRequesterIdAndIdAddresseeId(requester.id!!, 100L)

        assertThat(friendshipFound).isEmpty
    }

    companion object {
        private val REQUESTER_NICKNAME: String = "requester"
        private val ADDRESSEE_NICKNAME: String = "addressee"
    }
}